reactor-core/src/main/java/reactor/core/publisher/DirectProcessor.java

```diff
@@ -147,14 +158,26 @@ public Emission emitComplete() {
 	}
 
 	@Override
-	public Emission emitError(Throwable t) {
+	public void onError(Throwable throwable) {
+		emitError(throwable);
+	}
+
+	@Override
+	public void emitError(Throwable error) {
+		Emission result = tryEmitError(error);
+		if (result == Emission.FAIL_TERMINATED) {
```

Comment:
while migrating some of r-kafka code, I had the following in mind:
what if r-core adds more "negative" `Emission` cases but I only handle one.

Hence the idea:
Add `Emission#isSuccessful`, so that it can be checked in places like this one.

WDYT?

Reply:
there's `hasEmitted()` and `hasFailed()`. Maybe I should rename `hasEmitted()` to `hasSucceeded()`? not sure what you mean by "checked in places like this one" though... are you suggesting this piece of code should check something more?

Reply:
Ah, cool, so we can use `hasEmitted()` here (although renaming it to `hasSucceeded()` would be nice).
BTW isn't `hasFailed() == !hasEmitted()`? Why both?

re "can be checked in places like this one":
sorry, I meant this method is a good example where such boolean could be handy, and there are more methods like this one (in this PR but also in r-kafka, for example), hence "the flag that can be checked in places like this one"

Reply:
yeah, I think it was just added because it reads better. And most likely, `hasFailed()` is going to be the more interesting one to check...

I've opened a separate issue to track these

Reply:
added a commit for #2329 in the latest rebase (with the `orThrow` split out of the original tryEmit commit)

Reply:
Either I don't see it or don't get it :D 

Anyways, shouldn't it be:
```suggestion
		if (result.hasFailed()) {
```
?

Reply:
no. for an emitError, cancelled or overflown states don't warrant specific action, whereas an attempt at terminating twice is illegal and usually notified to the `onErrorDropped`

Reply:
ok 👍